tools/gcp-ips/main.go

```diff
@@ -0,0 +1,336 @@
+	outputDir = "output"
+)
+
+// projectResources stores the slices of addresses and instances for one project.
+// References:
+// https://godoc.org/google.golang.org/api/compute/v1#AddressAggregatedList
+// https://godoc.org/google.golang.org/api/compute/v1#InstanceAggregatedList
+//
+// Here we make the assumption that addresses and instances together will give
+// us all of the internal IPs used in a network. If this is not true,
+// projectResources should be expanded to include the missing resources, and
+// then make the appropriate API call in getResources to get the aggregated list
+type projectResources struct {
+	Project                string
+	AddressAggregatedList  *compute.AddressAggregatedList
+	InstanceAggregatedList *compute.InstanceAggregatedList
+}
+
+// addressInfo holds the fields that we care about in our output table.
+type addressInfo struct {
+	Project string
+	IP      string
+	Status  string
+	Subnet  string
+	User    string
+}
+
+// initClient initialize the Compute API client.
+func initClient() *compute.Service {
+	ctx := context.Background()
+
+	client, err := google.DefaultClient(ctx, compute.ComputeScope)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	computeService, err := compute.New(client)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	return computeService
+}
+
+// getServiceProjects returns a list of service projects for a given host project.
+func getServiceProjects(hostProject string, service *compute.Service) (*compute.ProjectsGetXpnResources, error) {
+	log.Printf("Looking for service projects in %s", hostProject)
+
+	res, err := service.Projects.GetXpnResources(hostProject).Do()
+
+	if err != nil {
+		log.Printf("Error getting service projects for %s: %v", hostProject, err)
+	}
+
+	return res, err
+}
+
+// getResources returns the addresses and instances for a project.
+func getResources(project string, service *compute.Service) *projectResources {
+	log.Printf("Looking for addresses and instances in %s", project)
+
+	addressAggregatedList, err := service.Addresses.AggregatedList(project).Do()
+
+	if err != nil {
+		log.Printf("Error getting addresses for %s: %v", project, err)
+	}
+
+	instanceAggregatedList, err := service.Instances.AggregatedList(project).Do()
+	if err != nil {
+		log.Printf("Error getting instances for %s: %v", project, err)
+	}
+
+	return &projectResources{
+		Project:                project,
+		AddressAggregatedList:  addressAggregatedList,
+		InstanceAggregatedList: instanceAggregatedList,
+	}
+}
+
+// getAllResources returns addresses and instances for all service projects
+// attached to a host project.
+func getAllResources(hostProject string, service *compute.Service) []*projectResources {
+	res, err := getServiceProjects(hostProject, service)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ch := make(chan *projectResources)
+	var wg sync.WaitGroup
+
+	// For each project, use a goroutine to get the resources for that project.
```

Comment:
There should be 2 API calls per service project, and there is a soft limit of 100 service projects per shared VPC. You're right though, there theoretically could be accounts with thousands of service projects per shared VPC, but I feel like it won't be an issue for most account with <100 service projects per shared VPC. On the other hand, it would be good to make this work to handle all cases. 

I have just reimplemented this section using a semaphore. I will start taking a look at adding retries with backoff + jitter. Thanks!